core/templates/services/html-length.service.ts

```diff
@@ -16,85 +16,142 @@
     * The above tags constitutes for non text nodes.
     */
-  nonTextTags = ['oppia-noninteractive-math', 'oppia-noninteractive-image'];
-
-  /**
-    * ['li','blockquote','em','strong', 'p', 'a']
-    * The above tags serve as the tags for text content. Blockquote,
-    * em, a, strong occur as descendants of p tag. li tags are
-    * descendants of ol/ul tags.
-    */
-  textTags = ['p', 'ul', 'ol'];
+  nonTextTags = [
+    'oppia-noninteractive-math',
+    'oppia-noninteractive-image',
+    'oppia-noninteractive-link',
+    'oppia-noninteractive-collapsible',
+    'oppia-noninteractive-video',
+    'oppia-noninteractive-tabs',
+    'oppia-noninteractive-skillreview'];
 
-  computeHtmlLengthInWords(htmlString: string): number {
+  computeHtmlLength(
+      htmlString: string,
+      calculationType: CalculationType): number {
     if (!htmlString) {
       this.loggerService.error('Empty string was passed to compute length');
       return 0;
     }
+
     const sanitizedHtml = this.sanitizer.sanitize(
       SecurityContext.HTML, htmlString) as string;
-    let totalWeight = this.calculateBaselineLength(sanitizedHtml);
 
     // Identify custom tags using regex on the original HTML string.
-    const customTagRegex = /<oppia-noninteractive-(?:math|image)[^>]*>/g;
-    const customTags = htmlString.match(customTagRegex);
+    const customTags = htmlString.match(CUSTOM_TAG_REGEX);
+    let customTagsLength = (customTags) ? customTags.length : 0;
+
+    let totalLength = this.calculateBaselineLength(
+      sanitizedHtml, calculationType, customTagsLength);
 
     if (customTags) {
       for (const customTag of customTags) {
-        totalWeight += this.getWeightForNonTextNodes(customTag);
+        totalLength += this.getWeightForNonTextNodes(
+          customTag, calculationType);
       }
     }
-    return totalWeight;
+    return totalLength;
   }
 
-  private calculateBaselineLength(sanitizedHtml: string): number {
+  calculateBaselineLength(
+      sanitizedHtml: string,
+      calculationType: CalculationType, CustomTagsLength: number): number {
     let domparser = new DOMParser();
-    let dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+    let dom: Document;
+    try {
+      dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+      if (dom.body.children.length === 0 && !CustomTagsLength) {
+        throw new Error(
+          'No HTML tags found. Ensure ' +
+          'that a valid string that includes HTML tags is provided.');
+      }
+    } catch (error) {
+      throw new Error(
+        'Failed to parse HTML string.' +
+        ' Ensure that a valid string that includes HTML tags is provided.');
+    }
     let totalWeight = 0;
     for (let tag of Array.from(dom.body.children)) {
-      const ltag = tag.tagName.toLowerCase();
-      if (this.textTags.includes(ltag)) {
-        totalWeight += this.getWeightForTextNodes(tag as HTMLElement);
-      }
+      const textContent = tag.textContent || '';
```

Comment:
You mentioned that this is due to strictNullChecks. Can you give a clearer explanation about that? You were talking about replacing with `tag!.textContent` but I don't understand why this solves the problem because this seems to confuse the issue being with `tag = null` vs `tag.textContent = null`.

Reply:
OK, the key reference here is https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent (good find). It looks like null is possible in some very special cases that won't apply here. That said, for simplicity, I suggest leaving the code as you have it, and add a comment above the line pointing to that reference to explain why we are guarding against null here.

Reply:
Done!